### PR TITLE
Add result builder to decl blocks

### PR DESCRIPTION
**Pull request by kimdv**
Hi @ahoppen 👋

I have not forgotten you, just had a busy at work and then summer vacation!
And European Championship in football where Denmark did it very good ⚽️
But I'll get some time to look at somethings again. 

From #291
This PR aims to A `MemberDeclBlock` should be directly creatable from a `MemberDeclListBuilder` result builder. This would probably take some hardcoding of `MemberDeclBlock` in the convenience initializers where the initialization of a `MemberDeclBlock` with a `MemberDeclListBuilder` is another convenience added (besides the current ones of: syntax collection through result builder, tokens by string literals and defaulting of tokens)

This will make it possible to make things like

```
StructDecl("Foo") {
    StructDecl("Bar")
}
```

I will next week start too look into it again, and catch up.
I will properly ask some questions

### CODE REVIEW

**Comment by ahoppen** [comment]
Cool. Thanks! And there’s no need to apologize, you are working on this in your free time after all. 😉 

Let me know when the code is ready for review or if you have any questions. I just took a look at the code and it still appears to be in an incomplete state to me.

**Comment by ahoppen** [comment]
@kimdv What’s the state of this PR? Is it ready for review yet or still work in progress?

**Comment by kimdv** [comment]
@ahoppen I think it is in a state where it is ready for review. 

I have rebased with latest master

**Review by ahoppen — COMMENTED** [review]
I just looked through the changes and thought it was a little over-complicated to only generate the `ExpressibleAs` protocols for types that are children of syntax collections – why not have them for all syntax nodes as they might also come in handy later? I also thought it was a bit of a shame that `ExprBuildable` didn’t conform to `ExpressibleAsExprBuildable`. 

I couldn’t resist the temptation to play around with your design myself and think it’s easiest to just implement the `ExpressibleAs` protocols in `Buildables.swift.gyb`. That way we can also conform protocols like `ExprBuildable` to their corresponding `ExpressibleAs` protocols. I pushed my draft idea to https://github.com/ahoppen/swift-syntax/commit/d6d4f12596451fda73b19cca58ecb55c26ee8122. What do you think?

**Comment by kimdv** [comment]
> I just looked through the changes and thought it was a little over-complicated to only generate the `ExpressibleAs` protocols for types that are children of syntax collections – why not have them for all syntax nodes as they might also come in handy later?

 Sure we can generate for them all. Did that initially, but went away because I wouldn't introduce a lot of protocols that wasn't used. Will add them again.

> I also thought it was a bit of a shame that `ExprBuildable` didn’t conform to `ExpressibleAsExprBuildable`.

My thought was just as we have a `ExprBuildable` protocol it would be a bit duplicated to have a `ExpressibleAsExprBuildable`?
But maybe there is some scenarios I haven't thought of.


> I couldn’t resist the temptation to play around with your design myself and think it’s easiest to just implement the `ExpressibleAs` protocols in `Buildables.swift.gyb`. That way we can also conform protocols like `ExprBuildable` to their corresponding `ExpressibleAs` protocols. I pushed my draft idea to [ahoppen@d6d4f12](https://github.com/ahoppen/swift-syntax/commit/d6d4f12596451fda73b19cca58ecb55c26ee8122). What do you think?

So we should move everything from `ExpressibleBuildables.swift.gyb` to `Buildables.swift.gyb`.

**Comment by ahoppen** [comment]
> Sure we can generate for them all. Did that initially, but went away because I wouldn't introduce a lot of protocols that wasn't used. Will add them again.

My thought process was that we are exposing these protocols publicly and it’s hard to explain why certain syntax kinds have `ExpressibleAs` protocols and some don’t in case someone wants to use them in their own codebase. Generating all just eliminates that problem.

>  My thought was just as we have a `ExprBuildable` protocol it would be a bit duplicated to have a `ExpressibleAsExprBuildable`?
But maybe there is some scenarios I haven't thought of.

That’s a good point. I hadn’t thought of that. But I think that 
- if we have an `ExressibleAsExprBuildable` protocol, `ExprBuildable` should definitely conform to it (how is `ExprBuildable` not expressible as `ExprBuildable` 🤯) or
- we should not have the `ExpressibleAsExprBuildable` protocol at all. In that case, `ResultBuilders.swift.gyb` would need to check whether `element_type` has an `ExpressibleAs` protocol or
- `typealias ExpressibleAsExprBuildable = ExprBuildable` and add a comment that this type alias only exists to simplify code generation.

I don’t have a strong preference for either of these options.

> So we should move everything from `ExpressibleBuildables.swift.gyb` to `Buildables.swift.gyb`.

Not necessarily. I only did so to conform `ExprBuildable` to `ExpressibleAsExprBuildable` (and reuse some of the loops). If we decide that we don’t need `ExressibleAsExprBuildable` as a protocol, everything can live inside `ExpressibleBuildables.swift.gyb` again.

**Comment by kimdv** [comment]
@ahoppen I have applied your suggestions 👍

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                    CaseItem(pattern: EnumCasePattern(caseName: version))
+                }), statementsBuilder: {
+                    ReturnStmt(expression: StringLiteralExpr(semVer, closeQuote: .stringQuote.withTrailingTrivia(.newlines(1))))
+                })
+            }
+        })
+
+        let syntax = switchStmt.buildSyntax(format: Format())
+
+        XCTAssertEqual(syntax.description, """
```
Comment (kimdv) [inline Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:75]: I have added this test case so we also have something more than `struct` examples.

BUT is keeps failing, and I don't know why. 
The expected result and the actual result seems equal.
Maybe you can help me @ahoppen ?

Screenshot 2021-09-23 at 22 11 10
?

**Comment by ahoppen** [comment]
> After some thoughts I think it would be strange that `ExprBuildable` would be the only one with no `ExpressibleAs` as protocol.

It wouldn’t have been the only one. `DeclBuildable` also wouldn’t have had an `ExpressibleAs` protocol 😉 Still I agree it makes sense to add them for consistency.

File: Sources/SwiftSyntaxBuilder/StringLiteralExprConvenienceInitializers.swift [hunk 2]
```diff
@@ -13,14 +13,14 @@
 import SwiftSyntax
 
 extension StringLiteralExpr {
-  public init(_ value: String) {
+    public init(_ value: String, openQuote: TokenSyntax = .stringQuote, closeQuote: TokenSyntax = .stringQuote) {
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/StringLiteralExprConvenienceInitializers.swift:16]: Incorrect indentation?

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    func testExpressibleAsMemberDeclListItem() {
+        let myStruct = StructDecl(identifier: "MyStruct", members: MemberDeclBlock(membersBuilder: {
+            VariableDecl(.var, name: "myFirstVar", type: "Int")
+
+            VariableDecl(letOrVarKeyword: .let, bindingsBuilder: {
+                PatternBinding(pattern: IdentifierPattern(identifier: .identifier("mySecondLet")),
+                               typeAnnotation: TypeAnnotation(type: SimpleTypeIdentifier("CGFloat")))
+            })
+
+            MemberDeclListItem(decl: VariableDecl(letOrVarKeyword: .let, bindingsBuilder: {
```
Comment (ahoppen) [inline Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:16]: Do you want to explicitly test we can put a `MemberDeclListItem` here? If so, I think it would be good to add a comment explaining your intention.

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import XCTest
+import SwiftSyntax
+
+import SwiftSyntaxBuilder
+
+final class ExpressibleBuildablesTests: XCTestCase {
+    func testExpressibleAsMemberDeclListItem() {
+        let myStruct = StructDecl(identifier: "MyStruct", members: MemberDeclBlock(membersBuilder: {
+            VariableDecl(.var, name: "myFirstVar", type: "Int")
```
Comment (ahoppen) [inline Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:9]: This is really starting to look like you can generate real-world code with `SwiftSyntaxBuilder` 😍

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+            VariableDecl(letOrVarKeyword: .let, bindingsBuilder: {
```
Comment (ahoppen) [inline Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:11]: Do you want to test that both the `VariableDecl` convenience initializer and the `bindingsBuilder` initializer work here? If so, could you add a comment explaining your intention?

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            let mySecondLet: CGFloat
+            let myOtherVar: String
+            struct InnerStruct{
+            }
+        }
+        """)
+    }
+
+    func testExpressibleAsCodeBlockItem() {
+        let myCodeBlock = SourceFile(eofToken: SyntaxFactory.makeToken(.eof, presence: .present)) {
```
Comment (ahoppen) [inline Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:39]: Not part of this PR but would it make sense to add 
```swift
extension TokenSyntax {
  static let eof = SyntaxFactory.makeToken(.eof, presence: .present)
}
```
So we could improve this to 
```suggestion
        let myCodeBlock = SourceFile(eofToken: .eof) {
```
And possibly try to remove the need to specify the `eofToken` in the `SourceFile` initializer altogether.

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 1 above)
Comment (ahoppen) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:75]: The generated code contains trailing whiteface after the `:`. It’s a bit annoying that you can’t have trailing whitespace in the string literal you’re matching against 😬 Maybe trim all trailing whitespace from the generated string before comparing or replace all whitespace with `␣`?

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        return "3.0.0"
+        case .version_3_1:
+        return "3.1.0"
+
+        }
+        """)
+    }
+}
+
+extension ReturnStmt: ExpressibleAsCodeBlockItem {
```
Comment (ahoppen) [inline Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:91]: This should probably live in `SwiftSyntaxBuilder` and not the test suite, right?

**Review by ahoppen — COMMENTED** [review]
Looking very good to me. And the test cases really start to look like what I’ve always envisioned `SwiftSyntaxBuilder` to look like 🤩 

I’ve added some minor comments regarding the test cases inline.

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 3 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:16]: My intention is to make sure they can be combined and showcasing it.
I will add a comment

Comment (ahoppen) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:16]: Thats’s good! Just wanted to make sure we’re on the same page what should be tested here.

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 7 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:91]: Oh yes. Sorry! 
Was playing around. I will also make Statements implement the `ExpressibleAsCodeBlockItem`. 
Then people can add switch case, return, if and so on.

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 5 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:11]: Removing it. cannot remember why.

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 1 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift]: Replaced with `␣` it makes it more clean to read the test case I think. 😬

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 7 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift]: @ahoppen added `ExpressibleAsCodeBlockItem` protocol to `Stmt`

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 4 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:9]: Yes! It look really good and it gets more and more clean to read! 🎉

**Review by ahoppen — APPROVED** [review]
Nice. Looks good to me. 🚀

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by kimdv** [comment]
Rebased on latest master

File: Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift (hunk 6 above)
Comment (kimdv) [reply Tests/SwiftSyntaxBuilderTest/ExpressibleBuildablesTests.swift:34]: Added in #314

**Comment by ahoppen** [comment]
@swift-ci Please test
